docs/source/coding_guidelines.md

```diff
@@ -0,0 +1,213 @@
+  open `./docs/build/html/index.html`
+
+Example:
+
+```python
+class Foo:
+    """
+    This is the class description.
+
+    The following block will be pretty-printed by Sphinx. Note the space between >>> and the code!
+    
+    Usage example:
+        >>> from module import Foo
+        >>> foo = Foo(bar=1.23)
+    """
+
+    ANY_ATTRIBUTE = "what_ever."
+    """Document class attributes after the attribute."""
+
+    def __init__(self, bar: float = 0.5) -> None:
+        """
+        :param bar: This is a description for the constructor argument.
+            Long descriptions should be indented.
+        """
+        self.bar = bar
+
+    def method(self, arg: int) -> None:
+        """
+        Method description, followed by an empty line. Argument names like ``arg`` are rendered nicely
+        if enclosed in double backtick.
+
+        This method can raise a :exc:`ValueError`.
+        
+        :param arg: This is a description for the method argument.
+            Long descriptions should be indented.
+        """
+```
+
+## Testing
+
+### Why do we write unit tests?
+
+If you think that there are sensitive parts in your code, where hidden errors can impact the project outcome, please take your time to protect parts of your code. As the team grows, so does the number of  users of the repository, and we may not be able to monitor changes on a frequent basis, but your tests can ensure that the expected behavior is preserved. We are not making it strictly mandatory, however users shall take the liability of untested code and its potential downstream impact.
+
+We do not need to test every part of the code depending on the project constraints, e.g., quality and test-coverage vs learning and hypothesis-testing. However, we expect a justification in each PR for this decision and it should be reviewed by team members. Please regularly monitor the code coverage results that are posted as comments on each pull request.
+
+In particular, please do write unit tests for anything that affects training/results such as data pre-processing, losses, and metrics.
+
+### Testing Do's / Don'ts
+
+* DO write unit tests for each new function or class that you add.
+* DO extend unit tests for existing functions or classes if you change their core behaviour.
+* DO try your best to write unit tests that are fast. Very often, this can be done by reducing data size to a minimum. Also, it is helpful to avoid long-running integration tests, but try to test at the level of the smallest involved function.
+* DO ensure that your tests are designed in a way that they can pass on the local machine, even if they are relying on specific cloud features. If required, use `unittest.mock` to simulate the cloud features, and hence enable the tests to run successfully on your local machine.
+* DO run all unit tests on your dev machine before submitting your changes. The test suite is designed to pass completely also outside of cloud builds.
```

Comment:
Should we replace "builds" with "CI"? Assuming CI has been defined in the first usage.